### PR TITLE
use kubeconfigs listed in KUBECONFIG env var

### LINKED ISSUE #6240: merge multiple files listed in KUBECONFIG environment var
### Is your enhancement related to a problem? Please describe

One can provide kubectl with multiple kube config files via the `KUBECONFIG` env variable. Kubectl would merge those and let the user use them.
ex. 
* `export KUBECONFIG=${HOME}/.kube/config:${HOME}/.kube/minikube.yaml:${HOME}/.kube/sandbox.yaml`
* empty `~/.kube/config` with only `current-context` pointing to a context that's defined in further file
```yaml
apiVersion: v1
kind: Config
clusters: null
contexts: null
current-context: minikube
preferences: {}
users: null
```
* `~/.kube/minikube.yml`
```yaml
apiVersion: v1
kind: Config
clusters:
  - cluster:
      certificate-authority: /Users/adietish/.minikube/ca.crt
      extensions:
        - extension:
            last-update: Wed, 07 Aug 2024 11:05:40 CEST
            provider: minikube.sigs.k8s.io
            version: v1.33.1
          name: cluster_info
      server: https://127.0.0.1:39093
    name: minikube
contexts:
  - context:
      cluster: minikube
      extensions:
        - extension:
            last-update: Wed, 07 Aug 2024 11:05:40 CEST
            provider: minikube.sigs.k8s.io
            version: v1.33.1
          name: context_info
      namespace: default
      user: minikube
    name: minikube
users:
  - name: minikube
    user:
      client-certificate: /Users/adietish/.minikube/profiles/minikube/client.crt
      client-key: /Users/adietish/.minikube/profiles/minikube/client.key
```
* `~/.kube/sandbox.yml`
```yaml
apiVersion: v1
kind: Config
clusters:
  - cluster:
      server: https://XXXX.openshiftapps.com:6443
    name: api-sandbox-m3-1530-p1-openshiftapps-com:6443
users:
  - name: adietish/XXXX-openshiftapps-com:6443
    user:
      token: sha256~YYYYYYYY
contexts:
  - context:
      cluster: XXXX-openshiftapps-com:6443
      namespace: adietish-dev
      user: adietish/XXXX-openshiftapps-com:6443
    name: adietish-dev/XXXX-openshiftapps-com:6443/adietish 
```

In kubectl I can then have `config` pointing to minikube or sandbox, both would work. In kubernetes-client, only the 1st file is used which then fails because it is missing url, user, etc.

### Describe the solution you'd like

It would be great if kubernetes-client could merge these files into one config like kubectl does.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by adietish**
## Description

Fixes #6240 

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [x] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/main/CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [x] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/main/doc/CHEATSHEET.md) accordingly
 - [x] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [x] I tested my code in Kubernetes
 - [x] I tested my code in OpenShift

### CODE REVIEW

**Comment by rohanKanojia** [comment]
Could you please run `mvn spotless:apply` to fix style failures?

**Comment by adietish** [comment]
@manusa, @rohanKanojia: can you please have a look at my implementation and provide feedback? thanks!

File: kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/ConfigTest.java
Comment (rohanKanojia) [inline kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/ConfigTest.java:1243]: Could you please use `kubeconfig` string literal instead in all tests? In my opinion, it's better to use string literal as the test would fail if someone changes the constant value.

File: kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/Config.java
Comment (rohanKanojia) [inline kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/Config.java:607]:
```suggestion
    if (Utils.isNotNullOrEmpty(files)) {
```

File: kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/Config.java
Comment (rohanKanojia) [inline kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/Config.java:1804]: Do we want to have this file in this class?

File: kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/Config.java
Comment (rohanKanojia) [inline kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/Config.java:1816]: You can use lombok `@Getter` for these

**Comment by rohanKanojia** [comment]
> PS. `DefaultMockServerTest` seem to fail in github checks even though they're fine for me locally.

Yes, this is a known issue https://github.com/fabric8io/kubernetes-client/issues/6228 Marc had raised a PR to fix it but looks like it's not properly fixed it.

File: kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/Config.java
Comment (adietish) [reply kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/Config.java:1804]: @rohanKanojia we can move it out if you prefer.

**Comment by adietish** [comment]
build fails again on `DefaultMockServerWebSocketTest`:
```
Error:  Failures: 
Error:    DefaultMockServerWebSocketTest.andUpgradeToWebSocket, with configured events, should emit onClose when done:106->andUpgradeToWebSocket, with configured events, should emit onClose when done_closure4:-1->andUpgradeToWebSocket, with configured events, should emit onClose when done_closure4:107 Condition failed with Exception:
```

**Comment by rohanKanojia** [comment]
A unit test seems to be failing on windows :
```
[INFO] 
Error:  Errors: 
Error:    KubernetesClientImplTest.shouldInstantiateClientUsingSerializeDeserialize:132 � KubernetesClient An error has occurred.
[INFO] 
Error:  Tests run: 240, Failures: 0, Errors: 1, Skipped: 6
```

**Comment by adietish** [comment]
@rohanKanojia fixed it, the new class `KubeConfigFile` needed jackson annotations to allow deserialization.

**Comment by adietish** [comment]
@rohanKanojia: fails again on `DefaultMockServerWebSocketTest`. Please review, thanks!

File: kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/internal/KubeConfigUtilsTest.java
Comment (rohanKanojia) [inline kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/internal/KubeConfigUtilsTest.java:168]: nit, I think these three tests can be made parameterized.

File: kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/internal/KubeConfigUtils.java
Comment (adietish) [reply kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/internal/KubeConfigUtils.java:227]: @rohanKanojia:
The 2 ifs are bogus, but the adding extensions is expected. Within `Preferences` there are 2 instance variables:
* a boolean `colors`
* and a list of `Extensions`.
We need to copy those extensions from the source to the destination. Missing something?

File: kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/internal/KubeConfigUtilsTest.java
Comment (adietish) [reply kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/internal/KubeConfigUtilsTest.java:168]: @rohanKanojia not sure that we want this to be parametrized. The parametrized version that I just implemented is far less readable. See here:
```java
  @ParameterizedTest
  @MethodSource("hasAuthInfoNamed_arguments")
  void hasAuthInfoNamed(Config config, String authInfoName, Consumer<Boolean> consumer) {
    // given
    // when
    boolean hasIt = KubeConfigUtils.hasAuthInfoNamed(config, authInfoName);
    // then
    consumer.accept(hasIt);
  },

  static Stream<Arguments> hasAuthInfoNamed_arguments() {
    return Stream.of(
      // given config with authInfo, when getAuthInfoName with existing name, then should re
      Arguments.of(getTestKubeConfig(), "test/api-test-com:443", (Consumer<Boolean>) (hasIt -> assertThat(hasIt).isTrue())),
      Arguments.of(getTestKubeConfig(), "bogus", (Consumer<Boolean>) (hasIt -> assertThat(hasIt).isFalse())),
      Arguments.of(new ConfigBuilder().build(), "test/api-test-com:443", (Consumer<Boolean>) (hasIt -> assertThat(hasIt).isFalse()))
    );
  }
```

**Comment by adietish** [comment]
@rohanKanojia, @manusa: the negative sonar status is imho erroneous. It erroneously omits a null-check, assertions present in a test case (parametrized test), complains about deprecation and blames too many (required) parameters in `SundrioConfig` constructor.

**Comment by adietish** [comment]
@rohanKanojia: fails again on DefaultMockServerWebSocketTest.

**Comment by manusa** [comment]
> @rohanKanojia: fails again on DefaultMockServerWebSocketTest.

This is a known failure.

We'll try to merge this PR once we merge others that might create conflicts on the Config class but need to go first because we need them for a [6.13.4 release](https://github.com/fabric8io/kubernetes-client/issues/6253).

**Comment by adietish** [comment]
@rohanKanojia, @manusa: could we please progress on this PR, please as it is important for us at intellij-kubernetes?

**Comment by manusa** [comment]
> @rohanKanojia, @manusa: could we please progress on this PR, please as it is important for us at intellij-kubernetes?

Please check previous [comment](https://github.com/fabric8io/kubernetes-client/pull/6295#issuecomment-2357460603).

Note also that this one will only be available in v7.0.0 we can't backport this to 6.13.
If it's really urgent or critical we can try to do a 6.14, but that will delay even more the 7.0.0 release. If you do need the backport + 6.14 release, please say so in order to plan it accordingly.

**Comment by adietish** [comment]
@manusa: if possible I'd love this to be available in `6.14` but I can live with it only being available with `7.0.0`. The reasoning is that we'd need all our plugins to use the same kubernetes-client (limitations in the jetbrains platform). I assume bumping all these would be substantial work and delay our releases substantially (we're currently on 6.12).

**Comment by manusa** [comment]
> I assume bumping all these would be substantial work and delay our releases substantially (we're currently on 6.12).

I don't think this will be the case for v7.0.0 and your use-case. It should be fairly straightforward to bump to the next major and should not require additional work.

You can do a quick check by bumping to 7.0-SNAOSHOT, at this point I don't think this will even require any code changes for the plugins.

**Comment by adietish** [comment]
@manusa, @rohanKanojia: do you have any updates regarding this PR and the release of 7.0.0 please?

**Comment by manusa** [comment]
Hi @adietish 

We had a few hiccups with the new model generation (#6130) regarding the OpenShift types which have delayed us around 1 week or more. The problems are now mostly solved, so we should be able to proceed with the rest of tasks in #5778.

Regarding this PR I hope I can merge it by the end of this week so that you can at least try it in the 7.0-SNAPSHOT release.
Please take into account that this PR is dealing with one of the most critical and delicate classes and functionalities of the client, so we have to be extra-careful when merging and implementing any additional test to ensure that the client retains its original behavior while adding the new functionality.

**Comment by adietish** [comment]
Hi @manusa 

thx for your extensive explanations, makes perfect sense.

> Regarding this PR I hope I can merge it by the end of this week so that you can at least try it in the 7.0-SNAPSHOT release. Please take into account that this PR is dealing with one of the most critical and delicate classes and functionalities of the client, so we have to be extra-careful when merging and implementing any additional test to ensure that the client retains its original behavior while adding the new functionality.

Fully agree.
I already did some additional tests, maybe I can have one more look into it to see if there's more I can assert.

**Comment by manusa** [comment]
Hi @adietish 
Before merging this PR I'd like to add some changes on top.
Are you done with your work? Is it OK if I take over the PR and add a few commits on top of your branch?

**Comment by manusa** [comment]
> @manusa Yes I'm done and confident that stuff works as expected. I had the changes reviewed/tested in our plugin, working as expected for our usecases.
Go ahead and take it. Looking forward to have this merged :)

Cool, I'm taking it over now.
Let's see if we can merge it soon.

**Review by manusa — APPROVED** [review]
LGTM, thx!

**Pull request merged**
